### PR TITLE
Make sure internal semaphore release is protected.

### PR DESCRIPTION
**Pull request by simon-shopify**
A code change moved `sem_meta_unlock` outside of the protected scope. This PR brings it back inside.

### CODE REVIEW

**Comment by dalehamel** [comment]
> A code change moved sem_meta_unlock outside of the protected scope

@simon-shopify what was the change?

**Comment by simon-shopify** [comment]
> @simon-shopify what was the change?

I think that was #120.

**Comment by dalehamel** [comment]
@simon-shopify i'm going on vacation in an hour, don't have time to give you a proper review. 

Seems sane to me.

**Comment by simon-shopify** [comment]
> @simon-shopify i'm going on vacation in an hour, don't have time to give you a proper review.

No worries, I'll ask someone else. Enjoy!

**Comment by csfrancis** [comment]
@simon-shopify could you add a test for this?

**Comment by simon-shopify** [comment]
> @simon-shopify could you add a test for this?

I'd like to, but I'm not sure how to do it. Is there a way to mock syscalls from Ruby? The issue is that it configures the ticket right after getting it so I don't have a chance to mess with it directly to trigger an error condition.

File: ext/semian/tickets.c [hunk 1]
```diff
@@ -2,71 +2,65 @@
...
+update_ticket_count(int sem_id, int count);
 
 static int
 calculate_quota_tickets(int sem_id, double quota);
 
 // Must be called with the semaphore meta lock already acquired
-void
-configure_tickets(int sem_id, int tickets, double quota)
+VALUE
+configure_tickets(VALUE value)
```
Comment (sirupsen) [inline ext/semian/tickets.c]: Why did you need to change the signature?

File: ext/semian/sysv_semaphores.c [hunk 2]
```diff
@@ -63,9 +63,23 @@ initialize_semaphore_set(semian_resource_t* res, const char* id_str, long permis
...
+  int state = 0;
   sem_meta_lock(res->sem_id); // Sets otime for the first time by acquiring the sem lock
-  configure_tickets(res->sem_id, tickets,  quota);
+
+  configure_tickets_args_t configure_tickets_args = (configure_tickets_args_t){
+    .sem_id = res->sem_id,
+    .tickets = tickets,
+    .quota = quota,
+  };
+  rb_protect(
```
Comment (sirupsen) [inline ext/semian/sysv_semaphores.c]: [As far as I can gather](https://silverhammermba.github.io/emberb/c/#rb_protect) this is required because `configure_tickets` might raise, which has to be done within a `rb_protect` call to not segfault the VM?

**Comment by simon-shopify** [comment]
> I don't see anything wrong, but can you provide some more context? How did this come up for us? How did you pin it down to this problem?

All of the resource semaphores are used (this is common in our job servers). We still use a number of tickets proportional to the number of workers with rescue. This causes a lot of thrashing on the semaphore limits on every deploy. In particular, in this case, a worker wanted to reduce the number of tickets because it was getting shut down.
* It grabbed the internal semaphore.
* Timed out as it tried to reduce the count.
* Raised.

Since the internal semaphore was not protected, it was never released, so the count could not be further adjusted.

File: ext/semian/tickets.c (hunk 1 above)
Comment (simon-shopify) [inline ext/semian/tickets.c]: `rb_protect` can only call functions with the exact signature.

File: ext/semian/sysv_semaphores.c (hunk 2 above)
Comment (simon-shopify) [inline ext/semian/sysv_semaphores.c]: It is not to avoid a segfault. See the explanation below.

**Comment by sirupsen** [comment]
What happens to that process that's trying to grab the internal lock and decrease the ticket count? Does it raise and abort abnormally now?

**Comment by simon-shopify** [comment]
> What happens to that process that's trying to grab the internal lock and decrease the ticket count? Does it raise and abort abnormally with an exception?

Yes, in addition to leaving the semaphore locked.

**Comment by simon-shopify** [comment]
I added a test.

**Comment by sirupsen** [comment]
> Yes, in addition to leaving the semaphore locked.

Isn't it dangerous to keep the semaphore locked? What will unlock it? Why are we OK with raising and exiting when we failed to decrement, which is likely not super uncommon?

**Comment by simon-shopify** [comment]
> Isn't it dangerous to keep the semaphore locked? What will unlock it? 

Sorry I was unclear. The bug is that the semaphore remains locked. This PR fixes it so it unlocks if there is a timeout.

> Why are we OK with raising and exiting when we failed to decrement, which is likely not super uncommon?

I don't think we are OK with it, but what is the alternative? I see it as graceful degradation.

**Comment by sirupsen** [comment]
> I don't think we are OK with it, but what is the alternative? I see it as graceful degradation.

The alternative would seem to be to log, instead of raise. This doesn't really seem exceptional

**Comment by simon-shopify** [comment]
> The alternative would seem to be to log, instead of raise. This doesn't really seem exceptional

Yes that could work as well. What do you think @csfrancis?

Note that this would not remove the need to `rb_protect` the semaphore since there are other reason it can raise.

**Comment by sirupsen** [comment]
> Note that this would not remove the need to rb_protect the semaphore since there are other reason it can raise.

👍 Mostly concerned with common, non-exceptional exceptions being raised and aborting normal shutdown, potentially leading to other issues down the line.

**Comment by csfrancis** [comment]
> Yes that could work as well. What do you think @csfrancis?

I'm a bit torn. On one hand I like raising because if this happens, the Semian ticket count is **not** configured the way the client expects. But on the other hand, subsequent calls to acquire the resource _should_ eventually end up reconfiguring the ticket count.

I'd prefer to leave it as-is, but if you want to change it to log that's fine with me as well.

**Comment by simon-shopify** [comment]
Here is what I suggest:
* Merge this PR like this to fix the regression, but keep the existing behaviour.
* Open a second PR to turn the exception into a log when there is contention on the internal lock.
* Make a release.

File: test/protected_resource_test.rb [hunk 3]
```diff
@@ -112,4 +112,43 @@ def test_responds_to_name_when_bulkhead_or_circuit_breaker_disabled
     assert_equal :no_bulkhead, Semian[:no_bulkhead].name
     assert_equal :no_circuit_breaker, Semian[:no_circuit_breaker].name
   end
+
+  def test_gracefully_fails_when_unable_to_decrement_ticket_count
```
Comment (csfrancis) [inline test/protected_resource_test.rb]: Very nice test!

Comment (simon-shopify) [inline test/protected_resource_test.rb]: I'm pretty satisfied with it, the only downside is that the timeout for the internal semaphore is 5s and not configurable, so it does take 5s.

**Comment by csfrancis** [comment]
> Merge this PR like this to fix the regression, but keep the existing behaviour.

I would do this and cut a release now. Changing the behaviour to log is not as important as fixing this bug.
